cpp/src/arrow/dataset/filter.cc

```diff
@@ -931,6 +931,22 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
   return null();
 }
 
+Result<Datum> CastOrDictionaryEncode(const Datum& arr,
```

Comment:
This doesn't seem to be static nor in the anonymous namespace?
(ditto for other private functions in this file?)

Reply:
https://issues.apache.org/jira/browse/ARROW-9492